test/ardop/test_ARDOPC.c

```diff
@@ -0,0 +1,131 @@
+	decompressed[8] = '\0';
+	assert_string_equal(decompressed, expect);
+}
+
+/*
+ * Check that the given `call`sign (with optional SSID) has the given
+ * compressed `hexstr` representation in the ARDOP protocol. The given
+ * `hexstr` must be 12 characters of hexadecimal like "b908e1b2c010"
+ */
+static void assert_callsign_wireline_is(const char *call, const char *hexstr)
+{
+	UCHAR compressed[6];
+	char compressed_hex[2*sizeof(compressed) + 1];
+	compressed_hex[2 * sizeof(compressed)] = '\0';
+
+	// check string must be exactly 12 characters long
+	assert_int_equal(strlen(hexstr), 2 * sizeof(compressed));
+
+	CompressCallsign(call, compressed);
+
+	for (size_t i = 0; i < sizeof(compressed); ++i) {
+		snprintf(&compressed_hex[2*i], 3, "%02x", (unsigned int)compressed[i]);
+	}
+	assert_string_equal(hexstr, compressed_hex);
+}
+
+/*
+ * Verify that the given `call_in` callsign (with optional SSID) becomes
+ * `call_out` after a round-trip through the callsign compress/decompress
+ * algorithm. This verifies that the callsign can be sent and received.
+ */
+static void assert_callsign_inout(const char *call_in, const char *call_out)
+{
+	char out[11];
+	UCHAR compressed[6];
+
+	CompressCallsign(call_in, compressed);
+	DeCompressCallsign(compressed, out, sizeof(out));
+	assert_string_equal(call_out, out);
+}
+
+/* test string-packing compression and decompression */
+static void test_ascii_6bit(void **state)
+{
+	(void)state; /* unused */
+
+	assert_6bit_inout("        ", "        ");
+	assert_6bit_inout("AZ09-_  ", "AZ09-_  ");
+	assert_6bit_inout("A      Z", "A      Z");
+	assert_6bit_inout("HI\0\0\0\0\0U", "HI      ");
+	assert_6bit_inout("abcxyz  ", "ABCXYZ  ");
+}
+
+/* test callsign wireline representation */
+static void test_compress_callsign(void **state)
+{
+	(void)state; /* unused */
+
+	// six character + SSID values are taken from direct
+	// invocation of CompressCallsign() in ardopcf eab1f3165a30.
+	assert_callsign_wireline_is("A1A", "851840000010");
+	assert_callsign_wireline_is("A1A-1", "851840000011");
+	assert_callsign_wireline_is("N0CALL", "b908e1b2c010");
+	assert_callsign_wireline_is("N0CALL-A", "b908e1b2c021");
+	assert_callsign_wireline_is("N0CALL-Z", "b908e1b2c03a");
+
+	// non-canonical representation
+	assert_callsign_wireline_is("N0CALL-0", "b908e1b2c010");
+}
+
+/* test callsign round-trip through the protocol */
+static void test_decompress_callsign(void **state)
+{
+	(void)state; /* unused */
+
+	assert_callsign_inout("A1A", "A1A");
+	assert_callsign_inout("A1A-1", "A1A-1");
+	assert_callsign_inout("N0CALL", "N0CALL");
+	assert_callsign_inout("N0CALL-A", "N0CALL-A");
+	assert_callsign_inout("N0CALL-Z", "N0CALL-Z");
+	assert_callsign_inout("LONGCAL-15", "LONGCAL-15");
+
+	// non-canonical representation
+	assert_callsign_inout("N0CALL-0", "N0CALL");
+
+	// too long callsign
+	assert_callsign_inout("CAFECAFE", "CAFECAF");
+
+	// overflow
+	assert_callsign_inout("MUCHTOOLON-G", "MUCHTOO");
```

Comment:
I agree that this may not be idea behavior, but it is better than it was.  Perhaps at some point we'll change it.  However, for now, this test provides a reminder of what behavior is implemented and will call attention if it is unintentionally changed.